File: servers/go/nexusview.go
package main

import (
	"github.com/vkuznet/gonexus/gonexus"
)

// EntriesResp is the /entries response body.
type EntriesResp struct {
	Key       string      `json:"key"`
	RootClass string      `json:"root_class"`
	Entries   []NodeBrief `json:"entries"`
}

// NodeBrief describes a direct child without descending into it.
type NodeBrief struct {
	Name  string `json:"name"`
	Type  string `json:"type"` // "group" or "field"
	Class string `json:"class,omitempty"`
	Dtype string `json:"dtype,omitempty"`
	Shape []int  `json:"shape,omitempty"`
	Units string `json:"units,omitempty"`
}

// GroupResp is the /group response body.
type GroupResp struct {
	Key      string                 `json:"key"`
	Path     string                 `json:"path"`
	Class    string                 `json:"class"`
	Attrs    map[string]interface{} `json:"attrs,omitempty"`
	Children []NodeBrief            `json:"children"`
}

// FieldResp is the /field response body. Values is always present but may
// be a truncated slice of the full array; see Total/Truncated.
type FieldResp struct {
	Key       string                 `json:"key"`
	Path      string                 `json:"path"`
	Dtype     string                 `json:"dtype"`
	Shape     []int                  `json:"shape,omitempty"`
	Units     string                 `json:"units,omitempty"`
	Attrs     map[string]interface{} `json:"attrs,omitempty"`
	Offset    int                    `json:"offset"`
	Count     int                    `json:"count"`
	Total     int                    `json:"total"`
	Truncated bool                   `json:"truncated"`
	Values    interface{}            `json:"values"`
}

func attrsMap(a *gonexus.AttrSet) map[string]interface{} {
	if a == nil || a.Len() == 0 {
		return nil
	}
	out := make(map[string]interface{}, a.Len())
	for _, k := range a.Keys() {
		v, ok := a.Get(k)
		if ok {
			out[k] = v.Value
		}
	}
	return out
}

func briefOf(name string, obj gonexus.NXobject) NodeBrief {
	switch o := obj.(type) {
	case *gonexus.NXfield:
		return NodeBrief{
			Name:  name,
			Type:  "field",
			Dtype: o.Dtype,
			Shape: o.Shape,
			Units: o.Units(),
		}
	case *gonexus.NXgroup:
		return NodeBrief{
			Name:  name,
			Type:  "group",
			Class: o.NXClass(),
		}
	default:
		return NodeBrief{Name: name, Type: obj.NXClass()}
	}
}

// buildEntries lists the NXentry groups directly under root (root is
// itself the NXroot group returned by gonexus.Load).
func buildEntries(key string, root *gonexus.NXgroup) *EntriesResp {
	resp := &EntriesResp{Key: key, RootClass: root.NXClass()}
	for _, e := range root.Entries() {
		resp.Entries = append(resp.Entries, briefOf(e.Name, e.Object))
	}
	return resp
}

// resolveGroupOrRoot looks up path under root, defaulting to root itself
// when path is empty, and asserts the result is a group.
func resolveGroupOrRoot(root *gonexus.NXgroup, path string) (*gonexus.NXgroup, error) {
	if path == "" || path == "/" {
		return root, nil
	}
	g, err := root.GetGroup(path)
	if err != nil {
		return nil, notFound("group not found at path %q: %v", path, err)
	}
	return g, nil
}

func buildGroup(key, path string, root *gonexus.NXgroup) (*GroupResp, error) {
	g, err := resolveGroupOrRoot(root, path)
	if err != nil {
		return nil, err
	}
	resp := &GroupResp{
		Key:   key,
		Path:  path,
		Class: g.NXClass(),
		Attrs: attrsMap(g.NXAttrs()),
	}
	for _, e := range g.Entries() {
		resp.Children = append(resp.Children, briefOf(e.Name, e.Object))
	}
	return resp, nil
}

func buildField(key, path string, root *gonexus.NXgroup, offset, limit int) (*FieldResp, error) {
	if path == "" {
		return nil, badRequest("missing required query parameter 'path'")
	}
	f, err := root.GetField(path)
	if err != nil {
		return nil, notFound("field not found at path %q: %v", path, err)
	}

	resp := &FieldResp{
		Key:   key,
		Path:  path,
		Dtype: f.Dtype,
		Shape: f.Shape,
		Units: f.Units(),
		Attrs: attrsMap(f.NXAttrs()),
	}

	if len(f.Shape) == 0 {
		// Scalar field: offset/limit don't apply.
		resp.Values = f.Value
		resp.Count = 1
		resp.Total = 1
		return resp, nil
	}

	values, total, truncated, err := sliceValues(f.Value, offset, limit)
	if err != nil {
		return nil, badRequest("%v", err)
	}
	resp.Values = values
	resp.Offset = offset
	resp.Total = total
	resp.Truncated = truncated
	switch v := values.(type) {
	case []float64:
		resp.Count = len(v)
	case []int64:
		resp.Count = len(v)
	case []string:
		resp.Count = len(v)
	case []bool:
		resp.Count = len(v)
	}
	return resp, nil
}

// sliceValues applies [offset:offset+limit] to a flat slice value of any
// of the types gonexus.NXfield.Value may hold, returning the slice, the
// total element count, and whether the slice is shorter than the total
// (i.e. the response is truncated).
func sliceValues(value interface{}, offset, limit int) (interface{}, int, bool, error) {
	switch v := value.(type) {
	case []float64:
		return sliceGeneric(v, offset, limit)
	case []int64:
		return sliceGeneric(v, offset, limit)
	case []int32:
		return sliceGeneric(toInt64s(v), offset, limit)
	case []float32:
		return sliceGeneric(toFloat64s(v), offset, limit)
	case []string:
		return sliceGeneric(v, offset, limit)
	case []bool:
		return sliceGeneric(v, offset, limit)
	default:
		// Fall back to Float64()/Int64() conversion via the field itself
		// is not available here (we only have the raw value), so just
		// report it as-is, unsliced, rather than fail outright.
		return value, 1, false, nil
	}
}

func toInt64s(v []int32) []int64 {
	out := make([]int64, len(v))
	for i, x := range v {
		out[i] = int64(x)
	}
	return out
}

func toFloat64s(v []float32) []float64 {
	out := make([]float64, len(v))
	for i, x := range v {
		out[i] = float64(x)
	}
	return out
}

func sliceGeneric[T any](v []T, offset, limit int) ([]T, int, bool, error) {
	total := len(v)
	if offset < 0 {
		offset = 0
	}
	if offset > total {
		offset = total
	}
	end := total
	if limit > 0 && offset+limit < end {
		end = offset + limit
	}
	out := v[offset:end]
	truncated := end < total
	return out, total, truncated, nil
}

File: servers/py/nexusview.py
"""Build JSON/text views of a NeXus (HDF5) file using plain h5py, mirroring
../go/nexusview.go so both server implementations expose identical shapes.
"""
import numpy as np
import h5py

from s3store import bad_request, not_found


def _py_scalar(v):
    """Convert a numpy scalar/bytes value to a plain JSON-serializable
    Python value."""
    if isinstance(v, bytes):
        return v.decode("utf-8", errors="replace")
    if isinstance(v, np.generic):
        return v.item()
    return v


def _attrs_map(obj):
    if obj.attrs is None or len(obj.attrs) == 0:
        return None
    out = {}
    for k in obj.attrs:
        out[k] = _py_scalar(obj.attrs[k])
    return out


def _units(obj):
    u = obj.attrs.get("units")
    if u is None:
        return None
    return _py_scalar(u)


def _dtype_name(dset: h5py.Dataset) -> str:
    return str(dset.dtype)


def _brief_of(name, obj):
    if isinstance(obj, h5py.Group):
        cls = obj.attrs.get("NX_class")
        cls = _py_scalar(cls) if cls is not None else "NXcollection"
        return {"name": name, "type": "group", "class": cls}
    # Dataset
    return {
        "name": name,
        "type": "field",
        "dtype": _dtype_name(obj),
        "shape": list(obj.shape) if obj.shape else [],
        "units": _units(obj),
    }


def build_entries(key, f: h5py.File):
    root_class = f.attrs.get("NX_class")
    root_class = _py_scalar(root_class) if root_class is not None else "NXroot"
    entries = []
    for name, obj in f.items():
        if isinstance(obj, h5py.Group):
            entries.append(_brief_of(name, obj))
    return {"key": key, "root_class": root_class, "entries": entries}


def _resolve_group_or_root(f: h5py.File, path: str):
    if not path or path == "/":
        return f
    try:
        obj = f[path]
    except KeyError:
        raise not_found(f"group not found at path {path!r}")
    if not isinstance(obj, h5py.Group):
        raise not_found(f"path {path!r} is not a group")
    return obj


def build_group(key, path, f: h5py.File):
    g = _resolve_group_or_root(f, path)
    cls = g.attrs.get("NX_class")
    cls = _py_scalar(cls) if cls is not None else ("NXroot" if g.name == "/" else "NXcollection")
    children = [_brief_of(name, obj) for name, obj in g.items()]
    return {
        "key": key,
        "path": path,
        "class": cls,
        "attrs": _attrs_map(g),
        "children": children,
    }


def build_field(key, path, f: h5py.File, offset: int, limit: int):
    if not path:
        raise bad_request("missing required query parameter 'path'")
    try:
        dset = f[path]
    except KeyError:
        raise not_found(f"field not found at path {path!r}")
    if not isinstance(dset, h5py.Dataset):
        raise not_found(f"path {path!r} is not a field")

    resp = {
        "key": key,
        "path": path,
        "dtype": _dtype_name(dset),
        "shape": list(dset.shape) if dset.shape else [],
        "units": _units(dset),
        "attrs": _attrs_map(dset),
    }

    if dset.shape == ():
        # Scalar field.
        resp["values"] = _py_scalar(dset[()])
        resp["offset"] = 0
        resp["count"] = 1
        resp["total"] = 1
        resp["truncated"] = False
        return resp

    # 1-D fields (and the leading axis of N-D fields) can be sliced
    # directly via h5py, which performs a partial HDF5 read rather than
    # loading the whole dataset into memory first.
    total = dset.shape[0]
    offset = max(0, offset)
    offset = min(offset, total)
    end = total
    if limit and limit > 0 and offset + limit < end:
        end = offset + limit
    chunk = dset[offset:end]
    truncated = end < total

    resp["offset"] = offset
    resp["total"] = total
    resp["truncated"] = truncated
    resp["count"] = int(chunk.shape[0]) if hasattr(chunk, "shape") else len(chunk)
    resp["values"] = _to_jsonable(chunk)
    return resp


def _to_jsonable(arr):
    if isinstance(arr, np.ndarray):
        if arr.dtype.kind in ("S", "O", "U"):
            return [
                v.decode("utf-8", errors="replace") if isinstance(v, bytes) else str(v)
                for v in arr.tolist()
            ]
        return arr.tolist()
    return arr


# ---------------------------------------------------------------------------
# Plain-text tree view, in the same spirit as gonexus's NXgroup.Tree().
# ---------------------------------------------------------------------------

_WELL_KNOWN_SKIP = {"NX_class"}


def _fmt_attrs(obj, indent):
    lines = []
    for k in obj.attrs:
        if k in _WELL_KNOWN_SKIP:
            continue
        v = _py_scalar(obj.attrs[k])
        lines.append(f"{indent}@{k} = {v}")
    return lines


def _fmt_node(name, obj, indent, lines):
    if isinstance(obj, h5py.Group):
        cls = obj.attrs.get("NX_class")
        cls = _py_scalar(cls) if cls is not None else "NXcollection"
        lines.append(f"{indent}{name}:{cls}")
        child_indent = indent + "  "
        lines.extend(_fmt_attrs(obj, child_indent))
        for child_name, child_obj in obj.items():
            _fmt_node(child_name, child_obj, child_indent, lines)
    else:
        shape = "x".join(str(d) for d in obj.shape) if obj.shape else ""
        shape_str = f"({shape})" if shape else ""
        lines.append(f"{indent}{name} = {obj.dtype}{shape_str}")
        for l in _fmt_attrs(obj, indent + "  "):
            lines.append(l)


def build_tree_text(f: h5py.File) -> str:
    lines = []
    root_cls = f.attrs.get("NX_class")
    root_cls = _py_scalar(root_cls) if root_cls is not None else "NXroot"
    lines.append(f"root:{root_cls}")
    lines.extend(_fmt_attrs(f, "  "))
    for name, obj in f.items():
        _fmt_node(name, obj, "  ", lines)
    return "\n".join(lines)
